conda/cli/main_config.py

```diff
@@ -807,3 +831,130 @@ def execute_config(args, parser):
+
+    try:
+        hardlink_dir_contents(root_prefix_pkgs, dest)
+    except (NotImplementedError, OSError):
+        # Hardlinks are not supported on all platforms, or between different
+        # filesystems; fall back a simple recursive copy instead.
+        #
+        # If the hardlink already exists, it shouldn't point to the <root_prefix>/pkgs/
+        # unless that was done manually. However, it probably contains the same package
+        # as we want (again, unless something was done manually.)
+        copy_dir_contents(root_prefix_pkgs, dest)
+    except Exception as e:
+        raise CondaError(f"Failed to migrate {root_prefix_pkgs} to {dest}.") from e
+
+
+def migrate_envs(context, _config: dict):
+    """Migrate the envs/ directory from the root prefix to the user data directory.
+
+    Uses hardlinks if they are supported, otherwise fall back to just copying
+    the directory contents instead.
+
+    The (empty) path at `<root prefix>/envs` will be removed if the move is successful.
+
+    :param context: Current execution context
+    :param config: Configuration dict read from `.condarc`
+    """
+    from .. import CondaError
+    from ..base.constants import USER_DATA_ENVS
+    from ..cli.main_rename import rename
+    from ..core.prefix_data import PrefixData
+
+    logger = getLogger(__name__)
+
+    if context._envs_dirs:
+        raise CondaError(
+            "The conda configuration for 'envs_dirs' has been explicitly set and "
+            "cannot be migrated."
+        )
+
+    layout = context.pkg_env_layout
+    if layout != PkgEnvLayout.USER.value:
+        raise CondaError(
+            f"The conda configuration for pkg_env_layout is set to {layout}; "
+            "Unless this setting is changed with `conda config set pkg_env_layout user` "
+            "conda will continue to store environments in the root prefix by default."
+        )
+
+    root_prefix_envs = Path(context.root_prefix_envs)
+    if not root_prefix_envs.is_dir():
+        raise CondaError(
+            f"The root prefix ({context.root_prefix}) does not contain an `envs` "
+            "directory. No migration is necessary."
+        )
+
+    failures = {}
+    dest = Path(USER_DATA_ENVS)
+    for env in os.listdir(root_prefix_envs):
+        prefix = root_prefix_envs / env
+        if PrefixData(prefix).is_environment():
+            try:
+                rename(
+                    source=str(prefix),
+                    destination=str(dest / env),
+                    dry_run=False,
+                    force=False,
+                    quiet=True,
+                    json=False,
+                )
+            except Exception as e:
+                logger.warning(
+                    f"Failed to migrate environment at {env} to {dest}. Reason: {e}"
+                )
+                failures[env] = e
+
+    if failures:
+        raise CondaError(
+            f"Errors migrating the following environments: {list(failures)}"
```

Comment:
Let's offer manual migration steps using `conda rename` the user can take here. This will be necessary if the environment name is already taken in the new location.

Reply:
Sure, manual migration steps have been added.